src/pages/instances/actions/InstanceBulkActions.tsx

```diff
@@ -0,0 +1,111 @@
+import React, { FC, Fragment, useState } from "react";
+import { updateInstanceBulkAction } from "api/instances";
+import { useNotify } from "context/notify";
+import { useQueryClient } from "@tanstack/react-query";
+import { queryKeys } from "util/queryKeys";
+import ConfirmationForce from "components/ConfirmationForce";
+import { LxdInstance, LxdInstanceAction } from "types/instance";
+import {
+  actionSummary,
+  instanceActions,
+  pluralizeInstance,
+} from "util/instanceBulkActions";
+import InstanceBulkAction from "pages/instances/actions/InstanceBulkAction";
+
+interface Props {
+  instances: LxdInstance[];
+  onStart: () => void;
+  onFinish: () => void;
+}
+
+const InstanceBulkActions: FC<Props> = ({ instances, onStart, onFinish }) => {
+  const notify = useNotify();
+  const queryClient = useQueryClient();
+  const [activeAction, setActiveAction] = useState<LxdInstanceAction | null>(
+    null
+  );
+  const [isForce, setForce] = useState(false);
+
+  const handleAction = (desiredAction: LxdInstanceAction) => {
+    setActiveAction(desiredAction);
+    onStart();
+    const actions = instanceActions(instances, desiredAction);
+
+    updateInstanceBulkAction(actions, isForce)
+      .then(() => {
+        const summary = actionSummary(actions);
+
+        notify.success(
+          summary.map(({ count, action }) => (
+            <Fragment key={action}>
+              <b>{count}</b> {pluralizeInstance(count)} {action}.<br />
+            </Fragment>
+          ))
+        );
+        setForce(false);
```

Comment:
Shouldn't this go in the `finally` block?

Reply:
This is intentional, if it failed, you want to redo and then it should stay as it was.

Reply:
Well, it should stay as it was if you had a dialog left open to let the user retry. But since the dialog is gone, and you reset everything in the `finally` block, you should definitely reset this prop as well. Otherwise, given that it's a shared prop between actions, you'd potentially leak a previous state to subsequent, unrelated, bulk operations.

Reply:
Ah true, user might do something else next. I'll change it.